feat(RELEASE-400): adds check-data-keys task in release pipeline

seanconroy2021 opened the pull request:
This commit adds the check-data-keys task into the release pipelines to ensure validation of required data keys.
[Link](https://issues.redhat.com/browse/RELEASE-400)

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented:
/ok-to-test

seanconroy2021 commented on schema/dataKeys.json:
```diff
@@ -49,8 +57,7 @@
         "productName": {
           "type": "string",
-          "description": "The product name e.g. exampleproduct ",
-          "pattern": "^[a-z]+$"
```
`  /workspace/data/5c52c6c8-898a-4a85-93b2-9336e6601e52/data.json::$.fbc.productName: 'testProductName' does not match '^[a-z]+$'`
  
  Just removing for time being to get E2E to pass.... I would say we can remove it

mmalina replied:
Or maybe just relax it to allow capital letters? Or modify e2e to contain lowercase in the name? That should be easy. Maybe just this here? https://github.com/konflux-ci/e2e-tests/blob/fd26460f5d0035b0bffe699513c558be8588b209/tests/release/pipelines/fbc_release.go#L347

johnbieren replied:
I hit this with some of the advisory keys too. I am fine with either removing it (the pattern, not the description) or modifying like Martin said so long as it is not too restricting such that users are going to fail on this

seanconroy2021 replied:
Ok so I have allowed for the pattern to have capital letters

mmalina replied:
looks good to me

seanconroy2021 replied:
We also get this one `.fbc.productName: 'preGA-product' does not match '^[a-zA-Z]+$.'` I could add in to allow hyphen or just update e2e.

johnbieren replied:
Up to you. I don't know that we have the exact requirements on this field which is why I say we can just remove it, but no strong opinion

seanconroy2021 commented:
/ok-to-test

johnbieren commented on pipelines/rh-advisories/rh-advisories.yaml:
This is the same as `rh-push-to-registry-redhat-io` but + advisories, so I think it should also have the systems that one has

seanconroy2021 replied:
Added in mapping, pyxis and sign into `rh-advisories`

johnbieren commented on pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml:
```diff
@@ -113,6 +113,32 @@ spec:
+            - pyxis
+            - mapping
+            - sign
+            - cosignSecretName
```
I don't think this is required

seanconroy2021 replied:
Fixed

johnbieren commented on pipelines/rhtap-service-push/rhtap-service-push.yaml:
```diff
@@ -109,6 +109,31 @@ spec:
+          value: 
+            - mapping
+            - pyxis
+            - sign
```
I don't see signing tasks in this pipeline, but I do see slack stuff

seanconroy2021 replied:
I have removed the signing. I thought Slack is not required  [Slack](https://github.com/seanconroy2021/release-service-catalog/blob/003c505a2127ee4f506736d0b8d9eb9435199ff9/tasks/send-slack-notification/send-slack-notification.yaml#L53-L54). I could be wrong

johnbieren replied:
You're right

johnbieren reviewed (commented):
E2E results may be interesting on this one :joy:

seanconroy2021 commented:
/ok-to-test

mmalina commented on pipelines/fbc-release/README.md:
```diff
@@ -19,6 +19,9 @@ Tekton release pipeline to interact with FBC Pipeline
 | taskGitRevision                 | The revision in the taskGitUrl repo to be used                                                           | No        | -                                                               |
 
+## Changes in 4.1.0
+* Implement the `check-data-keys` task to validate the `data.json` file using the JSON schema.
```
I would suggest to use the verb "Add" instead of "Implement"  - because that's what being done, right? We're adding the new task. (This applies to all the others as well.)

seanconroy2021 replied:
Yeah, we are adding .. will change it  
Fixed

mmalina commented on pipelines/rh-push-to-registry-redhat-io/rh-push-to-registry-redhat-io.yaml:
```diff
@@ -113,6 +113,32 @@ spec:
+        - name: data
+          workspace: release-workspace
+      runAfter:
+        - collect-data
```
In rh-advisories, this has verify-enterprise-contract in runAfter. Not sure which one is more correct, but they should do the same - and if run after verify-enterprise-contract is correct, then perhaps it should be used in other pipelines that have it as well.

seanconroy2021 replied:
They all should be run after `collect-data`  ... Will update the `push-to-cdn`
done

mmalina replied:
My comment is about rh-advisories ;) Not sure what you mean by "will update push-to-cdn" - that's a task. Which pipeline are you talking about?

seanconroy2021 replied:
I'm confused as ` rh-advisories ` has populate-release-notes-images in its run after for check-data-leys.

johnbieren replied:
Yeah, and we need to keep it that way AFAIK because it checks some of the keys the populate-release-notes-images task adds

mmalina replied:
> we need to keep it that way AFAIK because it checks some of the keys the populate-release-notes-images task adds

You mean if we checked the keys after populate-release-notes-images, it might not pass? But we just discussed elsewhere that the schema does include the releaseNotes.content.images, right? So it should be fine?

But I'm not sure who was even proposing we do the check after populate-release-notes-images here, certainly not me. 

My original question still stands: rh-advisories has verify-enterprise-contract in runAfter. Do we want that, or do we want collect-data? (Of course if we decide to have it after populate-release-notes-images in rh-advisories, this question would no longer be relevant.)

johnbieren replied:
rh-advisories does not have that. Are you looking at the right pipeline? It has this
```
    - name: check-data-keys
      params:
        - name: dataPath
          value: "$(tasks.collect-data.results.data)"
        - name: schema
          value:  $(params.taskGitUrl)/raw/$(params.taskGitRevision)/schema/dataKeys.json
        - name: systems
          value:
            - releaseNotes
      taskRef:
        params:
          - name: url
            value: $(params.taskGitUrl)
          - name: revision
            value: $(params.taskGitRevision)
          - name: pathInRepo
            value: tasks/check-data-keys/check-data-keys.yaml
        resolver: git
      workspaces:
        - name: data
          workspace: release-workspace
      runAfter:
        - populate-release-notes-images
```

mmalina replied:
> Are you looking at the right pipeline?

Thanks. I'm not sure where I was looking. You're right it's not there. Sorry about the noise.

But I do see it in `push-disk-images-to-cdn` pipeline:
```
    - name: check-data-keys
      params:
        - name: dataPath
          value: "$(tasks.collect-data.results.data)"
        - name: systems
          value:
            - cdn
      taskRef:
        params:
          - name: url
            value: $(params.taskGitUrl)
          - name: revision
            value: $(params.taskGitRevision)
          - name: pathInRepo
            value: tasks/check-data-keys/check-data-keys.yaml
        resolver: git
      workspaces:
        - name: data
          workspace: release-workspace
      runAfter:
        - verify-enterprise-contract

```

Do we know why? Shouldn't it also run right after collect-data there?

seanconroy2021 replied:
I would imagine it should run right after the collect-data like the rest.

mmalina replied:
Yeah, I'd do that.

seanconroy2021 replied:
Changed it

mmalina commented on schema/dataKeys.json:
```diff
@@ -422,7 +432,6 @@
                     "type": "array",
                     "items": {
                       "type": "object",
-                      "additionalProperties": false,
```
What's the motivation for this change? I don't know what's allowed for `contentGateway`, but I guess there was a reason to not allow additional properties?

seanconroy2021 replied:
It was down to extra params being added that are not user provided).
[Slack Thread here](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1731934464613269?thread_ts=1731665042.843589&cid=C04PZ7H0VA8)

mmalina replied:
Ah, ok, so it was based on a discussion, then it's good.

johnbieren reviewed (dismissed):
lgtm, would be nice to get Leandro's input on the fbc keys before merging

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest

johnbieren commented:
```
                        "lastTransitionTime": "2024-12-10T15:03:18Z",
                        "message": "the step \"reduce\" in TaskRun \"managed-gld9m-reduce-snapshot\" failed to pull the image \"\". The pod errored with the message: \"Back-off pulling image \"quay.io/enterprise-contract/ec-cli@sha256:913c7dac3d41877b01835d2e55bcd970c6cdbf4944f8176e9e3de9548642a2b4\".\"",
                        "reason": "TaskRunImagePullFailed",
                        "status": "False",
                        "type": "Succeeded"
```

theflockers commented on pipelines/push-disk-images-to-cdn/push-disk-images-to-cdn.yaml:
```diff
@@ -4,7 +4,7 @@ kind: Pipeline
   name: push-disk-images-to-cdn
   labels:
-    app.kubernetes.io/version: "1.2.0"
+    app.kubernetes.io/version: "1.2.1"
```
If I am not mistaken, for new parameters you should update the minor  - or 1.3.0 instead.

seanconroy2021 replied:
Fixed

theflockers commented on schema/dataKeys.json:
```diff
@@ -560,17 +564,186 @@
+              "targetIndex",
+              "publishingCredentials",
+              "buildTimeoutSeconds",
+              "issueId",
```
we have 3 types of FBC release: normal, preGa and hotfix. Some are common, other are only for preGA and other for hotfix. If I understand correctly, would this make all FBC require these keys?

seanconroy2021 replied:
So this would make all of them required. Which, I am guessing, would not work then?

theflockers replied:
Yes. Unless there is a way to create some conditions, like if hotfix, then... if preGA, then... otherwise, we should make only the common tags required. Also, I forgot to mention that when "stagedIndex" is true, the targetIndex should be empty :D

theflockers replied:
The following is true:

Required *common* parameters:
```
      fromIndex
      requestUpdateTimeout 
      buildTimeoutSeconds
      publishingCredentials
      allowedPackages
```
required for *prod*  builds
```
all common
      targetIndex
```
required for *stagedIndex* builds
```
all common
      stagedIndex

``` 
required for *hotfix* builds
```
all common
      targetIndex
      hotfix
      issueId

```
 required for *preGA* builds
```
all common
      targetIndex
      preGA
      productName
      productVersion
```

seanconroy2021 replied:
Thanks for the breakdown, I should be able to make it work

theflockers replied:
I updated the comment as the targetIndex was missing for preGA.

seanconroy2021 replied:
Ok, so i will able to make preGA and hotfix work from using if preGA or hotfix is true to make .. required. I cant see a way to make the others required since they only have one key

theflockers replied:
I think this is fine. The two you mentioned, their functionality rely on having these fields set. In the other hand, the stagedIndex will work the same way having the targetIndex or not.

theflockers replied:
Ah, the prod build means that stagedIndex is not set. Maybe that; s the only thing that still worth it to check. If `stageIndex` is false (or unset), the targetIndex is required.

seanconroy2021 replied:
Fixed

theflockers commented on tasks/check-data-keys/check-data-keys.yaml:
```diff
@@ -4,7 +4,7 @@ kind: Task
   name: check-data-keys
   labels:
-    app.kubernetes.io/version: "1.0.0"
+    app.kubernetes.io/version: "1.1.0"
```
adding a -L requires only bumping the patch, so 1.0.1.

seanconroy2021 replied:
Fixed

johnbieren commented on tasks/check-data-keys/README.md:
```diff
@@ -13,7 +13,11 @@ Currently, `releaseNotes`, and `cdn` are the only supported systems.
 | dataPath | Path to the JSON string of the merged data to use       | No       |               |
 | systems  | The systems to check that all data keys are present for | Yes      | []            |
-| schema   | The URl to the schema                                   | Yes      | https://github.com/konflux-ci/release-service-catalog/blob/production/schema/dataKeys.json |
+| schema   | The URl to the schema                                   | Yes      | https://github.com/konflux-ci/release-service-catalog/blob/development/schema/dataKeys.json |
```
Since you're fixing the default, can you make it what it actually is in the yaml? `https://raw.githubusercontent.com/konflux-ci/release-service-catalog/refs/heads/development/schema/dataKeys.json`

seanconroy2021 replied:
will do fixed

johnbieren reviewed (dismissed):
lgtm, but let Leandro approve the fbc changes before merging please

theflockers reviewed (commented):
LGTM once squashed and the failing test is fixed.

theflockers commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 committed:
feat(RELEASE-400): adds check-data-keys task in release pipelines

This commit adds the check-data-keys task into the
release pipelines to ensure validation of required data keys..

Signed-off-by: Sean Conroy <sconroy@redhat.com>

seanconroy2021 commented:
/retest

johnbieren commented:
> > @seanconroy2021 are you looking into why it fails or just retesting? E2E has been relatively stable this week
> 
> I have been looking at the logs not getting any info on why it failing... it keeps passing in the evening

If it is sometimes passing, you can enable auto-merge so that it will merge as soon as it passes e2e (assuming it is rebased)

seanconroy2021 commented:
/retest

theflockers commented:
/retest

seanconroy2021 commented:
/retest

seanconroy2021 commented:
/retest
